File: learning-core/executable/dbtx/cpp/test_dbtx.cpp
#include <cassert>

#include "dbtx.cpp"

int main() {
    Store store{{{"a", 100}, {"b", 50}}};
    auto tx = begin(store);
    transfer(tx, "a", "b", 30);
    commit(store, std::move(tx));
    assert(store.balances.at("a") == 70);
    assert(store.balances.at("b") == 80);

    const auto before = store.balances;
    auto failed = begin(store);
    try {
        transfer(failed, "a", "missing", 10);
        assert(false);
    } catch (const std::out_of_range&) {
    }
    assert(store.balances == before);
}

File: learning-core/executable/dbtx/python/test_dbtx.py
import sqlite3
import pytest

from dbtx import init_db, transfer


def balances(conn):
    return dict(conn.execute("select id,balance from accounts"))


def test_transfer_commits_both_changes() -> None:
    conn = sqlite3.connect(":memory:")
    init_db(conn)
    transfer(conn, "a", "b", 30)
    assert balances(conn) == {"a": 70, "b": 80}


def test_failure_rolls_back_transaction() -> None:
    conn = sqlite3.connect(":memory:")
    init_db(conn)
    before = balances(conn)
    with pytest.raises(KeyError):
        transfer(conn, "a", "missing", 30)
    assert balances(conn) == before
